### ISSUE TITLE
`set -o pipefail` in subshell incorrectly scoped to whole script (when detecting SC2312 check-extra-masked-returns)

### ISSUE THREAD

**Issue by x11x** [koalaman/shellcheck#2582]
#### For bugs
- Rule Id:  SC2312
- My shellcheck version (`shellcheck --version` or "online"): 0.8.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [ ] https://www.shellcheck.net/ (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/bash
# shellcheck enable=check-extra-masked-returns

a () {
  echo a
  false
}

b () {
  ( set -o pipefail
    a | cat
  )
}

c () {
  a | cat
}

b; echo $?
c; echo $?

```

#### Here's what shellcheck currently says:
    No issues detected!

#### Here's what I wanted or expected to see:

    Line 16:
      a | cat
      ^-- SC2312 (info): Consider invoking this command separately to avoid masking its return value (or use '|| true' to ignore).

In function `b`, the warning is correctly suppressed by using `set -o pipefail` in a subshell. Options set in subshell only apply in the subshell.

However, function `c` still has the problem of a masked return. Running the script in bash shows expected output of:

    a
    1
    a
    0

 It seems as though the `set -o pipefail` in the subshell is applying globally when it should just be scoped to the subshell. If I remove the entire `b () { ... }` function definition, the SC2312 problem is found in the `c` function, as expected.

**Comment by koalaman** [comment]
You're right, this is indeed a limitation. ShellCheck looks for certain options like `set -e`, `shopt -o pipefail`, and `shopt -o lastpipe` globally and optimistically hope they always apply because it's hard (especially for anything predating DFA support) to know exactly which commands are affected by which options given the possible paths through the script.

It's not an easy fix, but this issue can be left open as documentation.

**Comment by x11x** [comment]
Thanks for clarifying
